File: app/models/webhook.rb
class Webhook < ApplicationRecord
  belongs_to :contact, optional: true

  # Callbacks
  before_validation :generate_idempotency_key, on: :create

  # Validations
  validates :source, :event_type, :received_at, presence: true
  validates :status, inclusion: { in: %w[pending processing processed failed] }
  validates :idempotency_key, uniqueness: true, allow_nil: false

  # Scopes
  scope :pending, -> { where(status: 'pending') }
  scope :processing, -> { where(status: 'processing') }
  scope :processed, -> { where(status: 'processed') }
  scope :failed, -> { where(status: 'failed') }
  scope :unprocessed, -> { where(status: %w[pending failed]) }
  scope :recent, -> { where('received_at >= ?', 24.hours.ago) }

  # Source scopes
  scope :trust_hub, -> { where(source: 'twilio_trust_hub') }
  scope :sms_status, -> { where(source: 'twilio_sms') }
  scope :voice_status, -> { where(source: 'twilio_voice') }

  # Processing
  def process!
    # Use pessimistic locking to prevent race conditions (TOCTOU vulnerability)
    # Multiple Sidekiq workers could attempt to process the same webhook simultaneously
    with_lock do
      return if status == 'processed'

      # Wrap in transaction to ensure atomicity of multi-step updates
      transaction do
        update!(status: 'processing')

        begin
          case source
          when 'twilio_trust_hub'
            process_trust_hub_webhook
          when 'twilio_sms'
            process_sms_webhook
          when 'twilio_voice'
            process_voice_webhook
          else
            raise "Unknown webhook source: #{source}"
          end

          update!(
            status: 'processed',
            processed_at: Time.current
          )
        rescue StandardError => e
          update!(
            status: 'failed',
            processing_error: e.message,
            retry_count: (retry_count || 0) + 1
          )
          Rails.logger.error "Webhook processing failed: #{e.message}"
          raise
        end
      end
    end
  end

  # Retry failed webhooks
  def self.retry_failed!
    failed.where('retry_count < ?', 3).find_each do |webhook|
      webhook.process!
    end
  end

  # Auto-process pending webhooks (called from background job)
  def self.process_pending!
    pending.order(received_at: :asc).limit(100).find_each do |webhook|
      webhook.process!
    rescue StandardError => e
      Rails.logger.error "Failed to process webhook #{webhook.id}: #{e.message}"
    end
  end

  # Class method to get ransackable attributes
  def self.ransackable_attributes(auth_object = nil)
    ["source", "event_type", "status", "received_at", "processed_at",
     "created_at", "updated_at", "contact_id", "external_id"]
  end

  def self.ransackable_associations(auth_object = nil)
    ["contact"]
  end

  private

  def generate_idempotency_key
    # Generate unique key from source + external_id for idempotency protection
    # This prevents replay attacks where same webhook is POSTed multiple times
    return if idempotency_key.present?

    if external_id.present?
      self.idempotency_key = "#{source}:#{external_id}"
    else
      # Fallback: Use hash of payload if external_id is missing
      # This handles edge cases where Twilio doesn't provide an ID
      payload_hash = Digest::SHA256.hexdigest(payload.to_json.to_s)[0..31]
      self.idempotency_key = "#{source}:hash:#{payload_hash}"
      Rails.logger.warn("Webhook created without external_id, using payload hash for idempotency")
    end
  end

  def process_trust_hub_webhook
    # Extract Trust Hub data from payload
    # Handle both snake_case (from our code) and PascalCase (from Twilio webhooks)
    trust_hub_sid = payload['customer_profile_sid'] || payload['CustomerProfileSid'] || external_id
    status = payload['status'] || payload['Status']

    return unless trust_hub_sid.present?

    # Find contact by Trust Hub SID
    contact = Contact.find_by(trust_hub_customer_profile_sid: trust_hub_sid)

    if contact
      # Update Trust Hub status
      contact.update!(
        trust_hub_status: status,
        trust_hub_verified: ['twilio-approved', 'compliant'].include?(status),
        trust_hub_verification_score: calculate_trust_hub_score(status),
        trust_hub_enriched_at: Time.current
      )

      # Associate webhook with contact
      update!(contact_id: contact.id)

      Rails.logger.info "Trust Hub webhook processed for contact #{contact.id}: #{status}"
    else
      Rails.logger.warn "No contact found for Trust Hub SID: #{trust_hub_sid}"
    end
  end

  def process_sms_webhook
    # Extract SMS status from payload
    message_sid = payload['MessageSid'] || external_id
    message_status = payload['MessageStatus']
    to_number = payload['To']

    return unless to_number.present?

    # Find contact by phone number (optimized single query with OR condition)
    contact = Contact.where('formatted_phone_number = ? OR raw_phone_number = ?', to_number, to_number).first

    if contact
      # Update SMS tracking
      case message_status
      when 'delivered'
        contact.increment!(:sms_delivered_count)
      when 'failed', 'undelivered'
        contact.increment!(:sms_failed_count)
      end

      contact.update!(last_engagement_at: Time.current)
      update!(contact_id: contact.id)

      Rails.logger.info "SMS webhook processed for contact #{contact.id}: #{message_status}"
    end
  end

  def process_voice_webhook
    # Extract voice call status from payload
    call_sid = payload['CallSid'] || external_id
    call_status = payload['CallStatus']
    to_number = payload['To']

    return unless to_number.present?

    # Find contact by phone number (optimized single query with OR condition)
    contact = Contact.where('formatted_phone_number = ? OR raw_phone_number = ?', to_number, to_number).first

    if contact
      # Update voice tracking
      case call_status
      when 'completed'
        if payload['AnsweredBy'] == 'human'
          contact.increment!(:voice_answered_count)
        else
          contact.increment!(:voice_voicemail_count)
        end
      end

      contact.update!(
        voice_last_called_at: Time.current,
        last_engagement_at: Time.current
      )
      update!(contact_id: contact.id)

      Rails.logger.info "Voice webhook processed for contact #{contact.id}: #{call_status}"
    end
  end

  def calculate_trust_hub_score(status)
    case status
    when 'draft' then 10
    when 'pending-review' then 50
    when 'in-review' then 60
    when 'twilio-rejected', 'rejected' then 0
    when 'twilio-approved' then 100
    when 'compliant' then 95
    else 0
    end
  end
end

File: laravel-app/app/Models/Webhook.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Support\Facades\DB;

class Webhook extends Model
{
    use HasFactory;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<string>
     */
    protected $fillable = [
        'contact_id',
        'source',
        'event_type',
        'external_id',
        'payload',
        'headers',
        'status',
        'processed_at',
        'processing_error',
        'retry_count',
        'received_at',
        'idempotency_key',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'payload' => 'array',
        'headers' => 'array',
        'processed_at' => 'datetime',
        'received_at' => 'datetime',
    ];

    /**
     * Boot method for model events
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($webhook) {
            if (empty($webhook->idempotency_key)) {
                $webhook->generateIdempotencyKey();
            }
        });
    }

    /**
     * Relationships
     */
    public function contact(): BelongsTo
    {
        return $this->belongsTo(Contact::class);
    }

    /**
     * Query Scopes
     */
    public function scopePending($query)
    {
        return $query->where('status', 'pending');
    }

    public function scopeProcessing($query)
    {
        return $query->where('status', 'processing');
    }

    public function scopeProcessed($query)
    {
        return $query->where('status', 'processed');
    }

    public function scopeFailed($query)
    {
        return $query->where('status', 'failed');
    }

    public function scopeUnprocessed($query)
    {
        return $query->whereIn('status', ['pending', 'failed']);
    }

    public function scopeRecent($query)
    {
        return $query->where('received_at', '>=', now()->subDay());
    }

    public function scopeTrustHub($query)
    {
        return $query->where('source', 'twilio_trust_hub');
    }

    public function scopeSmsStatus($query)
    {
        return $query->where('source', 'twilio_sms');
    }

    public function scopeVoiceStatus($query)
    {
        return $query->where('source', 'twilio_voice');
    }

    /**
     * Generate idempotency key from source + external_id
     */
    private function generateIdempotencyKey(): void
    {
        if (!empty($this->external_id)) {
            $this->idempotency_key = "{$this->source}:{$this->external_id}";
        } else {
            // Fallback: Use hash of payload if external_id is missing
            $payloadHash = substr(hash('sha256', json_encode($this->payload ?? [])), 0, 32);
            $this->idempotency_key = "{$this->source}:hash:{$payloadHash}";
            \Log::warning('Webhook created without external_id, using payload hash for idempotency');
        }
    }

    /**
     * Process this webhook
     */
    public function process(): void
    {
        DB::transaction(function () {
            // Lock row to prevent concurrent processing
            $this->lockForUpdate();

            if ($this->status === 'processed') {
                return;
            }

            $this->update(['status' => 'processing']);

            try {
                switch ($this->source) {
                    case 'twilio_trust_hub':
                        $this->processTrustHubWebhook();
                        break;
                    case 'twilio_sms':
                        $this->processSmsWebhook();
                        break;
                    case 'twilio_voice':
                        $this->processVoiceWebhook();
                        break;
                    default:
                        throw new \Exception("Unknown webhook source: {$this->source}");
                }

                $this->update([
                    'status' => 'processed',
                    'processed_at' => now(),
                ]);
            } catch (\Exception $e) {
                $this->update([
                    'status' => 'failed',
                    'processing_error' => $e->getMessage(),
                    'retry_count' => ($this->retry_count ?? 0) + 1,
                ]);
                \Log::error("Webhook processing failed: {$e->getMessage()}");
                throw $e;
            }
        });
    }

    /**
     * Process Trust Hub webhook
     */
    private function processTrustHubWebhook(): void
    {
        $trustHubSid = $this->payload['customer_profile_sid'] ?? 
                       $this->payload['CustomerProfileSid'] ?? 
                       $this->external_id;
        $status = $this->payload['status'] ?? $this->payload['Status'];

        if (empty($trustHubSid)) {
            return;
        }

        $contact = Contact::where('trust_hub_customer_profile_sid', $trustHubSid)->first();

        if ($contact) {
            $contact->update([
                'trust_hub_status' => $status,
                'trust_hub_verified' => in_array($status, ['twilio-approved', 'compliant']),
                'trust_hub_verification_score' => $this->calculateTrustHubScore($status),
                'trust_hub_enriched_at' => now(),
            ]);

            $this->update(['contact_id' => $contact->id]);

            \Log::info("Trust Hub webhook processed for contact {$contact->id}: {$status}");
        } else {
            \Log::warning("No contact found for Trust Hub SID: {$trustHubSid}");
        }
    }

    /**
     * Process SMS webhook
     */
    private function processSmsWebhook(): void
    {
        $toNumber = $this->payload['To'] ?? null;

        if (empty($toNumber)) {
            return;
        }

        $contact = Contact::where('formatted_phone_number', $toNumber)
                         ->orWhere('raw_phone_number', $toNumber)
                         ->first();

        if (!$contact) {
            return;
        }

        $messageStatus = $this->payload['MessageStatus'] ?? null;

        // Update SMS tracking using atomic operations
        switch ($messageStatus) {
            case 'delivered':
                $contact->increment('sms_delivered_count');
                break;
            case 'failed':
            case 'undelivered':
                $contact->increment('sms_failed_count');
                break;
        }

        $contact->update(['last_engagement_at' => now()]);
        $this->update(['contact_id' => $contact->id]);

        \Log::info("SMS webhook processed for contact {$contact->id}: {$messageStatus}");
    }

    /**
     * Process voice webhook
     */
    private function processVoiceWebhook(): void
    {
        $toNumber = $this->payload['To'] ?? null;

        if (empty($toNumber)) {
            return;
        }

        $contact = Contact::where('formatted_phone_number', $toNumber)
                         ->orWhere('raw_phone_number', $toNumber)
                         ->first();

        if (!$contact) {
            return;
        }

        $callStatus = $this->payload['CallStatus'] ?? null;

        // Update voice tracking using atomic operations
        if ($callStatus === 'completed') {
            if (($this->payload['AnsweredBy'] ?? null) === 'human') {
                $contact->increment('voice_answered_count');
            } else {
                $contact->increment('voice_voicemail_count');
            }
        }

        $contact->update([
            'voice_last_called_at' => now(),
            'last_engagement_at' => now(),
        ]);
        $this->update(['contact_id' => $contact->id]);

        \Log::info("Voice webhook processed for contact {$contact->id}: {$callStatus}");
    }

    /**
     * Calculate Trust Hub score based on status
     */
    private function calculateTrustHubScore(string $status): int
    {
        return match ($status) {
            'draft' => 10,
            'pending-review' => 50,
            'in-review' => 60,
            'twilio-rejected', 'rejected' => 0,
            'twilio-approved' => 100,
            'compliant' => 95,
            default => 0,
        };
    }

    /**
     * Retry failed webhooks (static method)
     */
    public static function retryFailed(): void
    {
        self::failed()
            ->where('retry_count', '<', 3)
            ->each(function ($webhook) {
                try {
                    $webhook->process();
                } catch (\Exception $e) {
                    \Log::error("Failed to retry webhook {$webhook->id}: {$e->getMessage()}");
                }
            });
    }

    /**
     * Process pending webhooks (static method)
     */
    public static function processPending(): void
    {
        self::pending()
            ->orderBy('received_at')
            ->limit(100)
            ->each(function ($webhook) {
                try {
                    $webhook->process();
                } catch (\Exception $e) {
                    \Log::error("Failed to process webhook {$webhook->id}: {$e->getMessage()}");
                }
            });
    }
}
